### PR TITLE
[SPARK-41107][PYTHON][INFRA][TESTS] Install memory-profiler in the CI

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Install [memory-profiler](https://pypi.org/project/memory-profiler/) in the CI in order to enable memory profiling tests.

### Why are the changes needed?
That's a sub-task of [SPARK-40281](https://issues.apache.org/jira/browse/SPARK-40281) Memory Profiler on Executors.

PySpark memory profiler depends on memory-profiler. The PR proposes to install memory-profiler in the CI to enable related tests.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing tests.

SPARK-40281

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

File: dev/infra/Dockerfile [hunk 1]
```diff
@@ -32,7 +32,7 @@ RUN $APT_INSTALL software-properties-common git libxml2-dev pkg-config curl wget
 RUN update-alternatives --set java /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java
 
 RUN curl -sS https://bootstrap.pypa.io/get-pip.py | python3.9
-RUN python3.9 -m pip install numpy pyarrow 'pandas<=1.5.1' scipy unittest-xml-reporting plotly>=4.8 sklearn 'mlflow>=1.0' coverage matplotlib openpyxl
+RUN python3.9 -m pip install numpy pyarrow 'pandas<=1.5.1' scipy unittest-xml-reporting plotly>=4.8 sklearn 'mlflow>=1.0' coverage matplotlib openpyxl memory-profiler
```
Comment (ueshin) [inline dev/infra/Dockerfile:35]: Shall we fix the version of `memory-profiler`, just in case?

File: python/pyspark/testing/utils.py [hunk 2]
```diff
@@ -28,6 +28,7 @@
 
 have_scipy = False
 have_numpy = False
+have_memory_profiler = False
```
Comment (ueshin) [inline python/pyspark/testing/utils.py:31]: Can't we reuse `pyspark.profiler.has_memory_profiler`?

File: dev/infra/Dockerfile (hunk 1 above)
Comment (xinrong-meng) [reply dev/infra/Dockerfile]: Good point, modified.

File: python/pyspark/testing/utils.py (hunk 2 above)
Comment (xinrong-meng) [reply python/pyspark/testing/utils.py]: Sounds good.

**Review by ueshin — APPROVED** [review]

**Comment by xinrong-meng** [comment]
Failed as below. Irrelevant to the PR. Retriggered tests.

```
[error] (unsafe / Test / compileIncremental) Compilation failed
[error] Total time: 799 s (13:19), completed Nov 12, 2022 12:06:53 AM
[error] running /__w/spark/spark/build/sbt -Phadoop-3 -Phadoop-cloud -Pyarn -Pkinesis-asl -Pmesos -Pdocker-integration-tests -Pspark-ganglia-lgpl -Phive -Pkubernetes -Phive-thriftserver Test/package streaming-kinesis-asl-assembly/assembly connect/assembly ; received return code 1
Error: Process completed with exit code 16.
```

File: dev/infra/Dockerfile [hunk 3]
```diff
@@ -32,7 +32,7 @@ RUN $APT_INSTALL software-properties-common git libxml2-dev pkg-config curl wget
 RUN update-alternatives --set java /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java
 
 RUN curl -sS https://bootstrap.pypa.io/get-pip.py | python3.9
-RUN python3.9 -m pip install numpy pyarrow 'pandas<=1.5.1' scipy unittest-xml-reporting plotly>=4.8 sklearn 'mlflow>=1.0' coverage matplotlib openpyxl
+RUN python3.9 -m pip install numpy pyarrow 'pandas<=1.5.1' scipy unittest-xml-reporting plotly>=4.8 sklearn 'mlflow>=1.0' coverage matplotlib openpyxl 'memory-profiler==0.60.0'
 
 RUN add-apt-repository ppa:pypy/ppa
```
Comment (HyukjinKwon) [inline dev/infra/Dockerfile:37]:
```
 > [ 8/26] RUN add-apt-repository ppa:pypy/ppa:
#14 523.3 Cannot add PPA: 'ppa:~pypy/ubuntu/ppa'.
#14 523.3 ERROR: '~pypy' user or team does not exist.
------
Dockerfile:37
--------------------
  35 |     RUN python3.9 -m pip install numpy pyarrow 'pandas<=1.5.1' scipy unittest-xml-reporting plotly>=4.8 sklearn 'mlflow>=1.0' coverage matplotlib openpyxl 'memory-profiler==0.60.0'
  36 |     
  37 | >>> RUN add-apt-repository ppa:pypy/ppa
  38 |     RUN apt update
  39 |     RUN $APT_INSTALL gfortran libopenblas-dev liblapack-dev
```

@Yikun do you have an idea why this fails?

Comment (Yikun) [reply dev/infra/Dockerfile:37]: This should be the flaky error when adding apt repository (I haven't saw this error before), I think retry can help.

Comment (xinrong-meng) [reply dev/infra/Dockerfile:37]: Thanks! I'll retry.

**Comment by xinrong-meng** [comment]
Do you happen to know if it's normal to fail so many times? @Yikun

**Comment by HyukjinKwon** [comment]
I think mlflow got upgraded together for some reasons.

**Comment by Yikun** [comment]
Yes, new version mlflow had some breaking change, you could first install memory-profile in the end of dockerfile(like connect).

I will find sometime today to fix the doctest for new mlflow, in a separate PR.

**Review by Yikun — APPROVED** [review]
LGTM, I think it's ready to go

**Pull request closed without merging**